### PR TITLE
HPCC-23234 Fix problems if KJ in CQ and stopped early.

### PR DESCRIPTION
**Pull request by jakesmith**
If a KJ is inside a CQ and it's stopped early (e.g. due to a
downstream CHOOSEN), then it can lead to spurious results,
because it is not reset fully on restart.

Also problems with the limiters not being reset, which effectively
meant that it had an unbound lookahead.
That could cause memory issues, probably more of an issue if
KJ is inside a LOOP where each execution would be expected to
create larger outputs than if inside a CQ.

NB: commit also includes some minor tidying (e.g. moving const variables together).

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp [hunk 1]
```diff
@@ -706,6 +704,17 @@ class CJoinGroupList
             head = next;
         }
     }
+public:
+    CJoinGroupList() { }
+    ~CJoinGroupList()
+    {
+        removeAll();
+    }
+    void clear()
```
Comment (ghalliday) [inline thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp]: The split looks sensible, but this function doesn't seem to be called.

Comment (ghalliday) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp]: I take it back, my search wasn't finding it.

Comment (jakesmith) [reply thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp]: called on [new] lines 3048-3049

**Comment by ghalliday** [comment]
@jakesmith change looks sensible.  Any chance of a test case?

**Comment by jakesmith** [comment]
@ghalliday I discovered these issues, whilst developing a fix for HPCC-23211.
I've got a test case for that in the PR (https://github.com/hpcc-systems/HPCC-Platform/pull/13238), that exposes and tests this issue, as part of testing the fix.

so if we can wait for a test case, until HPCC-23211 is ready?

**Comment by ghalliday** [comment]
@jakesmith as long as there will be one eventually that is fine by me.  Is 7.6.x the correct target?

**Comment by jakesmith** [comment]
>Is 7.6.x the correct target?

I think so, it hasn't actually been reported, or new.

**Pull request merged**